schema/shared/src/main/scala-2/zio/blocks/schema/DerivedOptics.scala

```diff
@@ -0,0 +1,251 @@
+package zio.blocks.schema
+
+import scala.language.experimental.macros
+import scala.language.dynamics
+import scala.reflect.macros.whitebox
+
+/**
+ * A trait that can be extended by companion objects to automatically derive
+ * optics (Lens for case class fields, Prism for sealed trait/enum variants).
+ *
+ * Usage:
+ * {{{
+ * case class Person(name: String, age: Int)
+ * object Person extends DerivedOptics[Person]
+ *
+ * // Access optics via the `optics` member:
+ * val nameLens: Lens[Person, String] = Person.optics.name
+ * val ageLens: Lens[Person, Int] = Person.optics.age
+ * }}}
+ *
+ * For sealed traits:
+ * {{{
+ * sealed trait Shape
+ * case class Circle(radius: Double) extends Shape
+ * case class Rectangle(width: Double, height: Double) extends Shape
+ * object Shape extends DerivedOptics[Shape]
+ *
+ * // Access prisms via the `optics` member:
+ * val circlePrism: Prism[Shape, Circle] = Shape.optics.circle
+ * }}}
+ *
+ * The optics object is cached to avoid recreation on every access.
+ *
+ * @tparam S
+ *   The type for which to derive optics
+ */
+trait DerivedOptics[S] {
+
+  /**
+   * Provides access to the derived optics for type S. For case classes, returns
+   * an object with Lens accessors for each field. For sealed traits, returns an
+   * object with Prism accessors for each variant.
+   *
+   * The returned object uses structural typing, so you get compile-time type
+   * checking and IDE completion for the accessor names.
+   */
+  def optics(implicit schema: Schema[S]): Any = macro DerivedOpticsMacros.opticsImpl[S]
+}
+
+/**
+ * A variant of DerivedOptics that prefixes all accessor names with underscore.
+ * This is useful when you want to avoid name collisions with existing methods
+ * in the companion object.
+ *
+ * Usage:
+ * {{{
+ * case class Person(name: String, age: Int)
+ * object Person extends DerivedOptics_[Person]
+ *
+ * // Access optics with underscore prefix:
+ * val nameLens: Lens[Person, String] = Person.optics._name
+ * val ageLens: Lens[Person, Int] = Person.optics._age
+ * }}}
+ */
+trait DerivedOptics_[S] {
+
+  /**
+   * Provides access to the derived optics for type S with underscore-prefixed
+   * accessor names.
+   */
+  def optics(implicit schema: Schema[S]): Any = macro DerivedOpticsMacros.opticsImplUnderscore[S]
+}
+
+object DerivedOptics {
+  private[schema] final class OpticsHolder(members: Map[String, Any]) extends scala.Dynamic {
+    def selectDynamic(name: String): Any =
+      members.getOrElse(name, throw new RuntimeException(s"No optic found for: $name"))
+  }
+}
+
+object DerivedOpticsMacros {
```

Comment:
To the maximum extent possible, this companion object and other pure implementation details should be hidden from Scala docs. That means `private[schema]` at minimum.